File: pkg-py/tests/test_remote_interpro.py
"""InterPro remote mode: live existence checks via the EBI InterPro API."""

import pytest

import biogate
import biogate._remote as remote
from biogate import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_url_builds_the_interpro_entry_endpoint():
    from biogate._registry import get_source

    url = remote._interpro_url(get_source("interpro"), "IPR000001")
    assert url == "https://www.ebi.ac.uk/interpro/api/entry/interpro/IPR000001"


def test_existing_entry_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biogate.check_id("IPR000001", source_db="interpro", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "IPR000001"
    assert res.suggestion is None


def test_absent_entry_is_invalid(monkeypatch):
    # The InterPro entry endpoint answers 204 for a well-formed but absent id.
    monkeypatch.setattr(remote, "_http_get", _stub(204))
    res = biogate.check_id("IPR999999", source_db="interpro", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biogate.check_id("ipr000001", source_db="interpro", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "IPR000001"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biogate.check_id("IPRABCDEF", source_db="interpro", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biogate.check_id("IPR000001", source_db="interpro", how="remote")
    path = remote._remote_cache_path("interpro", "interpro", "IPR000001")
    assert not path.is_file()

File: pkg-r/tests/testthat/test-remote-interpro.R
# Offline tests for interpro remote mode. Existence is checked against the EBI
# InterPro API; the biobouncer.remote_transport option replaces it.

.stub_interpro <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}

test_that("the url addresses the interpro entry endpoint", {
  source <- list(remote = list(interpro_db = "interpro"))
  url <- .remote_resolvers$interpro$url(source, "IPR000001")
  expect_identical(
    url,
    "https://www.ebi.ac.uk/interpro/api/entry/interpro/IPR000001"
  )
})

test_that("an existing entry is valid", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_interpro(200))
  res <- check_id("IPR000001", "interpro", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "IPR000001")
  expect_true(is.na(res$suggestion))
})

test_that("an absent entry is invalid", {
  # The InterPro entry endpoint answers 204 for a well-formed but absent id.
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_interpro(204))
  res <- check_id("IPR999999", "interpro", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_interpro(200))
  res <- check_id("ipr000001", "interpro", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "IPR000001")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biobouncer.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("IPRABCDEF", "interpro", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .stub_interpro(500))
  expect_error(
    check_id("IPR000001", "interpro", how = "remote"),
    class = "biobouncer_error_remote"
  )
  expect_false(
    file.exists(.remote_cache_path("interpro", "interpro", "IPR000001"))
  )
})
